datafusion/datasource/src/url.rs

```diff
@@ -242,33 +242,64 @@ impl ListingTableUrl {
     ) -> Result<BoxStream<'a, Result<ObjectMeta>>> {
         let exec_options = &ctx.config_options().execution;
         let ignore_subdirectory = exec_options.listing_table_ignore_subdirectory;
-        // If the prefix is a file, use a head request, otherwise list
-        let list = match self.is_collection() {
-            true => match ctx.runtime_env().cache_manager.get_list_files_cache() {
-                None => store.list(Some(&self.prefix)),
+
+        async fn list_with_cache<'b>(
+            ctx: &'b dyn Session,
+            store: &'b dyn ObjectStore,
+            prefix: &'b Path,
+        ) -> Result<BoxStream<'b, Result<ObjectMeta>>> {
+            match ctx.runtime_env().cache_manager.get_list_files_cache() {
+                None => Ok(store
+                    .list(Some(prefix))
+                    .map(|res| res.map_err(|e| DataFusionError::ObjectStore(Box::new(e))))
```

Comment:
What is the reason for the map and then map err? I think you can express the same thing like

```suggestion
                    .map_err(|e| DataFusionError::ObjectStore(Box::new(e)))
```